query/query.go

```diff
@@ -0,0 +1,169 @@
+package query
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/fabric8-services/fabric8-wit/errors"
+	"github.com/fabric8-services/fabric8-wit/gormsupport"
+	"github.com/fabric8-services/fabric8-wit/log"
+	"github.com/goadesign/goa"
+	"github.com/jinzhu/gorm"
+	uuid "github.com/satori/go.uuid"
+)
+
+// APIStringTypeQuery helps to avoid string literal
+const APIStringTypeQuery = "queries"
+
+// Query describes a single Label
+type Query struct {
+	gormsupport.Lifecycle
+	ID      uuid.UUID `sql:"type:uuid default uuid_generate_v4()" gorm:"primary_key"` // This is the ID PK field
+	SpaceID uuid.UUID `sql:"type:uuid"`
+	Creator uuid.UUID `sql:"type:uuid"`
+	Title   string
+	Fields  string
+}
+
+// GetETagData returns the field values to use to generate the ETag
+// func (q Query) GetETagData() []interface{} {
+// 	return []interface{}{q.ID, q.Version}
+// }
+
+// GetLastModified returns the last modification time
+func (q Query) GetLastModified() time.Time {
+	return q.UpdatedAt.Truncate(time.Second)
+}
+
+// TableName overrides the table name settings in Gorm to force a specific table name
+// in the database.
+func (q Query) TableName() string {
+	return "queries"
+}
+
+// Repository describes interactions with Labels
+type Repository interface {
+	Create(ctx context.Context, u *Query) error
+	List(ctx context.Context, spaceID uuid.UUID) ([]Query, error)
+	ListByCreator(ctx context.Context, spaceID uuid.UUID, creatorID uuid.UUID) ([]Query, error)
+	Load(ctx context.Context, queryID uuid.UUID, spaceID uuid.UUID) (*Query, error)
+	Delete(ctx context.Context, ID uuid.UUID) error
+}
+
+// NewQueryRepository creates a new storage type.
+func NewQueryRepository(db *gorm.DB) Repository {
+	return &GormQueryRepository{db: db}
+}
+
+// GormQueryRepository is the implementation of the storage interface for Labels.
+type GormQueryRepository struct {
+	db *gorm.DB
+}
+
+// GetETagData returns the field values to use to generate the ETag
+func (q Query) GetETagData() []interface{} {
+	return []interface{}{q.ID, strconv.FormatInt(q.UpdatedAt.Unix(), 10)}
+}
+
+// QueryTableName constant that holds table name of Labels
+const QueryTableName = "queries"
+
+// Create a new label
+func (r *GormQueryRepository) Create(ctx context.Context, q *Query) error {
```

Comment:
Typo: "label" should be "query". Also, I'm missing a call to `ParseFilterString` in this function. It blindly accepts anything that is valid JSON it seems.

Reply:
You are right but I wanted repository unaware of application logic, so I am using  `ParseFilterString` in the controller before save to validate syntax & keys.

Reply:
The query itself is not handled in application logic but also on repository level. Therefore your intention does not hold true. I think we should do data validation on repo level.